### PR TITLE
USDZExporter: Store only unique geometry and reference it

### LINKED ISSUE #21748: USDExporter: browser tab crashes during export of complex geometry
**Describe the bug**

When I try to export my three.js objects to USDZ on iOS -> I have a browser's tab crash because of memory limit.
In my case I have many high-detailed objects which are using same geometry.
Final USDA will store all of them as different geometries for every Xform object.

Crashes only on mobile but desktop also uses huge amount of memory:
- Takes ~1+ GB RAM to do a conversion. 
- Output USDZ is ~200 MB. #21590 

**Expected behavior**

We could reference same geometries. So it will use much smaller amount of memory and disk space.

**Platform:**

 - Platform: Any
 - Three.js version: [dev, r128]

### PR DESCRIPTION
**Pull request by qeeqez**
Fixed #21748 

**Description**

Now we are storing only unique geometries in different geometry.usd file.
model.usda have only references to geometries (meshes).

Also, geometry do not store material references. Xforms from model.usda now do it. 
It could help us with storing of many identical objects but with different materials binded

### CODE REVIEW

**Comment by mrdoob** [comment]
Thanks!

File: examples/js/exporters/USDZExporter.js [hunk 1]
```diff
@@ -13,23 +17,37 @@
...
 				}
 
 			} );
 			output += buildMaterials( materials );
 			output += buildTextures( textures );
-			const files = {
-				'model.usda': fflate.strToU8( output )
-			};
+			files[ modelFileName ] = fflate.strToU8( output );
+			output = null;
```
Comment (mrdoob) [inline examples/js/exporters/USDZExporter.js:49]: This is to make the browser to deallocate memory, right?

Comment (qeeqez) [reply examples/js/exporters/USDZExporter.js:49]: Yes. Most of the memory will be allocated during geometry file creation. So we are clearing "output" string right before it.

**Comment by mrdoob** [comment]
> Now we are storing only unique geometries in different geometry.usd file.

Out of curiosity, why do we need to store them in a different file?

**Comment by qeeqez** [comment]
> > Now we are storing only unique geometries in different geometry.usd file.
> 
> Out of curiosity, why do we need to store them in a different file?

If we'll add geometry to the "output" - it will be instantiated in zero point. And we will see geometry itself in zero point + instantiated objects in correct places. 

We have multiple solutions:
1. Make geometry invisible. 
But overriding this invisible state in the instance do not work thus we could not see instances ¯\\\_(ツ)\_/¯
2. Store geometry in the first object's instance and then reference it. 
Good solution but we'll have some problems with overriding materials and etc.
3. **Store geometry in a different file**. This one is merged. 
Pixar's USD examples uses this approach in many places so I think it's a best solution.
   - [Some complex example](https://graphics.pixar.com/usd/docs/567231471.html)
   - [Simplier example](https://graphics.pixar.com/usd/docs/Referencing-Layers.html)
   p.s. looks like they are storing every geometry in a separate file: we could do the same so we'll further improve memory usage during file creation. But final USDZ will be a little bit bigger as every .usd file should have a header on the top.

Note: not all of the features of USDZ format are implemented or work in AR QuickLook. 
We could have some solutions working in the Xcode AR View but not in the AR Quick Look

**Comment by mrdoob** [comment]
> p.s. looks like they are storing every geometry in a separate file: we could do the same so we'll further improve memory usage during file creation.

Yes, I was thinking if that would be better. I found it weird when I tested this and I saw the geometry.usd file with all geometries on top of each other. Makes it hard to debug.

> But final USDZ will be a little bit bigger as every .usd file should have a header on the top.

I think it'd be worth it.

**Comment by qeeqez** [comment]
@mrdoob 
Question about file structure we want to have:
| Now  | Proposed |
| ------------- | ------------- |
| **model.usda**<br/>def Xform "Object_ID"<br/>references = @./geometry.usd@</Geometry_ID>  | **model.usda**<br/>def Xform "Object_ID"<br/>references = @./geometries/**Geometry_ID**.usd@</Geometry> |
| **geometry.usd**<br/>def "Geometry_ID"<br/>def Mesh "Geometry_ID"| **geometry_ID.usd**<br/>def "Geometry"<br/>def Mesh "Geometry"|

Basically my proposal - store geometry ID in file names (similar to current textures storage). And inside the file we will only use "Geometry" word without ID.

**Comment by mrdoob** [comment]
The proposed approach looks great to me 👍
